python/tweets/searcher.py

```diff
@@ -0,0 +1,97 @@
+# - json, for working with the tweets themselves (they are returned as json objects)
+# - pymongo, for connecting to a mongoDB mLab database where we'll store the tweets
+import tweepy, json, pymongo
+
+
+# Set up API - replace keys with your own: https://apps.twitter.com/
+CONSUMER_KEY = "<CONSUMER_KEY>"
+CONSUMER_SECRET = "<CONSUMER_SECRET>"
+ACCESS_KEY = "<ACCESS_KEY>"
+ACCESS_SECRET = "<ACCESS_TOKEN>"
+auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
+auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)
+api = tweepy.API(auth)
+
+
+# Bath, England place id for twitter api is "1db4f0a70fc5c9db", found using api.geo_search:
+# api.geo_search(query="Bath, England", granularity="city")[0]
+# We'll search for tweets from here
+search_query = "place:1db4f0a70fc5c9db"
+
+# Max tweets to grab in one go (just a big number)
+max_tweets = 1000000000
+# Set to max per query that API allows (100)
+tweets_per_query = 100
+
+# Connect to the mLab database
+client = pymongo.MongoClient("mongodb://shareduser:shareduser@ds123614.mlab.com:23614/bmlm-parking-tweets")
+db = client["bmlm-parking-tweets"]
+
+# If we already have some tweets, get the date of the most recent, and only look for tweets after that
+try:
+    most_recent = [tweet for tweet in db.tweets.find().sort("id", pymongo.DESCENDING)]
+    since_id = most_recent[0]["id"]
+except:
+    # Otherwise get all tweets that the API allows us to
+    since_id = None
+
+# Start from the most recent tweet matching the search query
+max_id = -1
+
+
+# Initialise a list to temporarily store the tweets
+tweets_to_add = []
+tweet_count = 0
+
+print("Downloading max {0} tweets".format(max_tweets))
+
+# Keep going until we hit max (or run out of tweets to get)
+while tweet_count < max_tweets:
+    
+    # Grab a block of tweets
+    try:
+        if (max_id <= 0):
+            if (not since_id):
+                new_tweets = api.search(q=search_query, count=tweets_per_query)
+            else:
+                new_tweets = api.search(q=search_query, count=tweets_per_query,
+                                        since_id=since_id)
+        else:
+            if (not since_id):
+                new_tweets = api.search(q=search_query, count=tweets_per_query,
+                                        max_id=str(max_id - 1))
+            else:
+                new_tweets = api.search(q=search_query, count=tweets_per_query,
+                                        max_id=str(max_id - 1),
+                                        since_id=since_id)
+        if not new_tweets:
+            print("No more tweets found")
+            break
+        
+        # Add this block of tweets to our list
+        tweets_to_add.extend([status._json for status in new_tweets])
+        tweet_count += len(new_tweets)
+        
+        print("Downloaded {0} tweets".format(tweet_count))
+        
+        # Set starting point for next block
+        max_id = new_tweets[-1].id
+        
+    except tweepy.TweepError as e:
```

Comment:
Not sure I see the benefit in this. Wouldn't the exception cancel stop the execution anyway, or am I missing something?

Reply:
You're right. It's because I've been practising try/except recently! Also it allows me to give (in theory) a more user-friendly error message:
```
$ python searcher.py
Downloading max 1000000000 tweets
Oh noooooo! The API rejected your auth keys - make sure they're set correctly!
No tweets to add :(
Downloaded 0 tweets, saved to mLab database
```
versus
```
$ python searcher.py
Downloading max 1000000000 tweets
Traceback (most recent call last):
  File "searcher.py", line 63, in <module>
    since_id=since_id)
  File "C:\Users\Owen\Anaconda3\lib\site-packages\tweepy\binder.py", line 245, in _call
    return method.execute()
  File "C:\Users\Owen\Anaconda3\lib\site-packages\tweepy\binder.py", line 229, in execute
    raise TweepError(error_msg, resp, api_code=api_error_code)
tweepy.error.TweepError: [{'code': 89, 'message': 'Invalid or expired token.'}]
```

Reply:
I'd argue the latter one is more user friendly, as our user is a programmer and it literally told me the line number where it crashed.

Reply:
Perhaps, perhaps... I am probably biased, because when I saw the Python error for the first time it took me an embarrassingly long time to figure out what the problem was. But you're right, we are programmers. (At least, I pretend to be.)

Reply:
I'll change this.